src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -2998,5 +2998,92 @@ public static bool IsFunctionPointerCallingConventionContext(this SyntaxTree syn
+        {
+            var possibleBaseListSyntax = targetToken.GetAncestor<BaseListSyntax>();
+            var possibleBaseTypeSyntax = targetToken.GetAncestor<BaseTypeSyntax>();
+
+            if (possibleBaseListSyntax?.Parent is not ClassDeclarationSyntax)
+            {
+                return false;
+            }
+
+            // We are in a non-written base type, e.g. "class C : $$"
+            if (possibleBaseTypeSyntax is null)
+            {
+                return targetToken.IsKind(SyntaxKind.ColonToken);
+            }
+
+            // 1. Current base type must be the first in the base list, because class inheritance is valid only at the first position in the base list
+            // 2. The only available target token type is a qualified name or an alias qualified name. If there is generic names or whatever else,
+            // then it is not directly inheritance context and our restrictions are not valid there
+            return possibleBaseListSyntax.Types[0].Equals(possibleBaseTypeSyntax) &&
+                   targetToken.Parent is QualifiedNameSyntax { Parent: BaseTypeSyntax or QualifiedNameSyntax { Parent: BaseTypeSyntax } }
+                                      or AliasQualifiedNameSyntax { Parent: BaseTypeSyntax or QualifiedNameSyntax { Parent: BaseTypeSyntax } };
+        }
+
+        /// <summary>
+        /// Computes <see cref="CSharpSyntaxContext.IsBaseEnumContext"/>
+        /// </summary>
+        public static bool IsBaseEnumContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            // We don't care about qualified names, generics or whatever else, because in base list of enum can only contain one item of byte, short etc.
+            return targetToken.GetAncestor<BaseListSyntax>() is { Parent: EnumDeclarationSyntax };
+        }
+
+        /// <summary>
+        /// Tells if in a position of a base list, where interface name is a valid completion.
+        /// For instatnce, this flag will be <see langword="true" /> for
+        /// <code>class C : $$</code>
+        /// or
+        /// <code>interface I : $$</code>
+        /// </summary>
+        public static bool IsBaseInterfaceContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            // The first part of condition checks for non-written base type, e.g. "class C : $$" or "class C : Base, $$"
+            // The second one checks quilified name cases, e.g. "class C : System.$$" or "class C : global::$$"
+            return (targetToken.IsKind(SyntaxKind.ColonToken, SyntaxKind.CommaToken) && targetToken.Parent is BaseListSyntax { Parent: not EnumDeclarationSyntax }) ||
+                    targetToken.Parent is QualifiedNameSyntax { Parent: BaseTypeSyntax or QualifiedNameSyntax { Parent: BaseTypeSyntax } }
+                                       or AliasQualifiedNameSyntax { Parent: BaseTypeSyntax or QualifiedNameSyntax { Parent: BaseTypeSyntax } };
```

Comment:
split into multiple checks.  the && and || mixing is def confusing.  also, i think the qualified name checks may not be accurate (they do not account for things on the rhs).